components/ShadowLayer/src/MDCShadowLayer.m

```diff
@@ -304,25 +334,87 @@ - (void)commonLayoutSublayers {
   _topShadow.bounds = bounds;
 
   if (_shadowMaskEnabled) {
-    _bottomShadow.mask = [self shadowLayerMaskForLayer:_bottomShadow];
-    _topShadow.mask = [self shadowLayerMaskForLayer:_topShadow];
+    [self configureShadowLayerMaskForLayer:_topShadowMask];
+    [self configureShadowLayerMaskForLayer:_bottomShadowMask];
   }
   // Enforce shadowPaths because otherwise no shadows can be drawn. If a shadowPath
   // is already set, use that, otherwise fallback to just a regular rect because path.
-  if (!_bottomShadow.shadowPath) {
+  if (!_bottomShadow.shadowPath || _shadowPathIsInvalid) {
     if (self.shadowPath) {
       _bottomShadow.shadowPath = self.shadowPath;
     } else {
       _bottomShadow.shadowPath = [self defaultShadowPath].CGPath;
     }
   }
-  if (!_topShadow.shadowPath) {
+  if (!_topShadow.shadowPath || _shadowPathIsInvalid) {
     if (self.shadowPath) {
       _topShadow.shadowPath = self.shadowPath;
     } else {
       _topShadow.shadowPath = [self defaultShadowPath].CGPath;
     }
   }
+  _shadowPathIsInvalid = NO;
+}
+
+@end
+
+@implementation MDCAnimatedShapeLayer
```

Comment:
You can remove this intermediary object if you implement CALayerDelegate on `MDCShadowLayer` and you move `actionForKey:`'s logic into `actionForLayer:forKey:`.

Reply:
Each of the layers would then set their layer delegate to the MDCShadowLayer instance and you'd avoid having to manage parent/child relationships (you'll still need to store a reference to the shadow layer in the pending animation instance though).

Reply:
Good to know. I'll give this a go.